examples/ubersum.py

```diff
@@ -0,0 +1,192 @@
+    """
+    key = 'gradient', equation, kwargs['batch_dims']
+    if key not in _CACHE:
+
+        # This wraps ubersum for jit compilation, but we will call backward on the result.
+        def _ubersum(*operands):
+            return ubersum(equation, *operands, **kwargs)
+
+        _CACHE[key] = torch.jit.trace(_ubersum, operands, check_trace=False)
+
+    # Run forward pass.
+    losses = _CACHE[key](*operands)
+
+    # Work around PyTorch 1.0.0 bug https://github.com/pytorch/pytorch/issues/14875
+    # whereby tuples of length 1 are unwrapped by the jit.
+    if not isinstance(losses, tuple):
+        losses = (losses,)
+
+    # Run backward pass.
+    grads = tuple(grad(loss, operands, retain_graph=True, allow_unused=True)
+                  for loss in losses)
+    return grads
+
+
+def _jit_adjoint(equation, *operands, **kwargs):
+    """
+    Runs ubersum in forward-backward mode using ``pyro.ops.adjoint``.
+
+    This simulates serving predictions from an undirected graphical model.
+    """
+    backend = kwargs.pop('backend', 'pyro.ops.einsum.torch_sample')
+    key = backend, equation, tuple(x.shape for x in operands), kwargs['batch_dims']
+    if key not in _CACHE:
+
+        # This wraps a complete adjoint algorithm call.
+        @ignore_jit_warnings()
+        def _forward_backward(*operands):
+            # First we request backward results on each input operand.
+            # This is the pyro.ops.adjoint equivalent of torch's .require_grad_().
+            for operand in operands:
+                require_backward(operand)
+
+            # Next we run the forward pass.
+            results = ubersum(equation, *operands, backend=backend, **kwargs)
+
+            # The we run a backward pass.
+            for result in results:
+                result._pyro_backward()
+
+            # Finally we retrieve results from the ._pyro_backward_result attribute
+            # that has been set on each input operand. If you only want results on a
+            # subset of operands, you can call require_backward() on only those.
+            results = []
+            for x in operands:
+                results.append(x._pyro_backward_result)
+                x._pyro_backward_result = None
+
+            return tuple(results)
+
+        _CACHE[key] = torch.jit.trace(_forward_backward, operands, check_trace=False)
+
+    return _CACHE[key](*operands)
+
+
+def jit_map(equation, *operands, **kwargs):
+    return _jit_adjoint(equation, *operands, backend='pyro.ops.einsum.torch_map', **kwargs)
+
+
+def jit_sample(equation, *operands, **kwargs):
+    return _jit_adjoint(equation, *operands, backend='pyro.ops.einsum.torch_sample', **kwargs)
+
+
+def jit_marginal(equation, *operands, **kwargs):
+    return _jit_adjoint(equation, *operands, backend='pyro.ops.einsum.torch_marginal', **kwargs)
+
+
+def time_fn(fn, equation, *operands, **kwargs):
+    iters = kwargs.pop('iters')
+    _CACHE.clear()  # Avoid memory leaks.
```

Comment:
Just to clarify from our discussion yesterday - this isn't just to avoid memory leaks, but to get the correctly compiled function for different plate sizes too, right?

Reply:
Clearning was only needed for correctness by `naive_ubersum` in our discussion yesterday. non-naive Ubersum never slices things up, and so still be correct for different sized tensors.